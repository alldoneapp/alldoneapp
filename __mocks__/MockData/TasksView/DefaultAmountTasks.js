export const DefaultAmountTasks = [
    { open: 0, pending: 0, done: 0 },
    { open: 0, pending: 0, done: 0 },
]
