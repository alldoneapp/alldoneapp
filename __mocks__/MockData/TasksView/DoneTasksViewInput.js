export const DoneTasksViewInput = [
    [
        //
        // Done tasks
        {
            id: '-LcRZz0eNJkx7QZn_gti2',
            done: true,
            name: 'later yo',
            userIds: ['C08CK8x1I5YS2lxVixuLHaF3SrA3'],
            hasStar: false,
            created: 1555192800000,
            creatorId: 'C08CK8x1I5YS2lxVixuLHaF3SrA3',
            dueDate: 1564735497020,
            completed: 1564735497020,
            isPrivate: false,
            parentId: null,
            subtaskIds: [],
            subtaskNames: [],
            recurrence: { type: 'never' },
            inReview: null,
            toReview: null,
            receivedFrom: {},
            normalTasks: []
        },
        {
            id: '-LcRZz0eNJkx7QZn_gti3',
            done: true,
            name: 'later yo',
            userIds: ['C08CK8x1I5YS2lxVixuLHaF3SrA3'],
            hasStar: false,
            created: 1555192800020,
            creatorId: 'C08CK8x1I5YS2lxVixuLHaF3SrA3',
            dueDate: 1564735497040,
            completed: 1564735497040,
            isPrivate: false,
            parentId: null,
            subtaskIds: [],
            subtaskNames: [],
            recurrence: { type: 'never' },
            inReview: null,
            toReview: null,
            receivedFrom: {},
            normalTasks: []
        },
    ],
    [
        {
            //
            // Done tasks
            id: '-LcRh9ABE_wbmsAhSUiN4',
            done: true,
            name: 'test unknown yo',
            userIds: ['C08CK8x1I5YS2lxVixuLHaF3SrA3'],
            hasStar: false,
            created: 1555192800000,
            creatorId: 'C08CK8x1I5YS2lxVixuLHaF3SrA3',
            dueDate: 1555274088568,
            completed: 1555274088568,
            isPrivate: false,
            parentId: null,
            subtaskIds: [],
            subtaskNames: [],
            recurrence: { type: 'never' },
            inReview: null,
            toReview: null,
            receivedFrom: {},
            normalTasks: []
        },
        {
            //
            // Done tasks
            id: '-LcRh9ABE_wbmsAhSUiN5',
            done: true,
            name: 'test unknown yo',
            userIds: ['C08CK8x1I5YS2lxVixuLHaF3SrA3'],
            hasStar: false,
            created: 1555192800000,
            creatorId: 'C08CK8x1I5YS2lxVixuLHaF3SrA3',
            dueDate: 1555274088568,
            completed: 1555274088568,
            isPrivate: false,
            parentId: null,
            subtaskIds: [],
            subtaskNames: [],
            recurrence: { type: 'never' },
            inReview: null,
            toReview: null,
            receivedFrom: {},
            normalTasks: []
        },
    ],
]
