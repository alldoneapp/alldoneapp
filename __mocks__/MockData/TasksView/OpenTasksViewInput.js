export const OpenTasksViewInput = [
    [
        {
            id: '-LcRVT6MEWlqGQRkE2xw',
            done: false,
            name: 'My new task',
            userIds: ['C08CK8x1I5YS2lxVixuLHaF3SrA3'],
            hasStar: false,
            created: 1555192800000,
            creatorId: 'C08CK8x1I5YS2lxVixuLHaF3SrA3',
            dueDate: 1555273253219,
            completed: 1555273253219,
            isPrivate: false,
            parentId: null,
            subtaskIds: [],
            subtaskNames: [],
            recurrence: { type: 'never' },
            inReview: null,
            toReview: null,
        },
        {
            id: '-LcRZz0eNJkx7QZn_gti',
            done: false,
            name: 'later yo',
            userIds: ['C08CK8x1I5YS2lxVixuLHaF3SrA3'],
            hasStar: false,
            created: 1555192800000,
            creatorId: 'C08CK8x1I5YS2lxVixuLHaF3SrA3',
            dueDate: 1564735497020,
            completed: 1564735497020,
            isPrivate: false,
            parentId: null,
            subtaskIds: [],
            subtaskNames: [],
            recurrence: { type: 'never' },
            inReview: null,
            toReview: null,
        },
    ],
    [
        {
            id: '-LcRh9ABE_wbmsAhSUiN',
            done: false,
            name: 'test unknown yo',
            userIds: ['C08CK8x1I5YS2lxVixuLHaF3SrA3'],
            hasStar: false,
            created: 1555192800000,
            creatorId: 'C08CK8x1I5YS2lxVixuLHaF3SrA3',
            dueDate: 1555274088568,
            completed: 1555274088568,
            isPrivate: false,
            parentId: null,
            subtaskIds: [],
            subtaskNames: [],
            recurrence: { type: 'never' },
            inReview: null,
            toReview: null,
        },
    ],
]
