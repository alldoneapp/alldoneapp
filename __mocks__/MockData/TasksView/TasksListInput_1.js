export const TasksOpen_1 = {
    id: '-Txyz_1',
    done: false,
    name: 'My new open task',
    userIds: ['-Uxyz_1'],
    hasStar: false,
    created: 1555192800000,
    creatorId: '-Uxyz_1',
    dueDate: 1555273253219,
    completed: 1555273253219,
    isPrivate: false,
    parentId: null,
    subtaskIds: [],
    subtaskNames: [],
    recurrence: { type: 'never' },
}

export const TasksPending_1 = {
    id: '-Txyz_2',
    done: false,
    name: 'My new pending task',
    userIds: ['-Uxyz_1', '-Uxyz_2'],
    hasStar: false,
    created: 1555192800000,
    creatorId: '-Uxyz_1',
    dueDate: 1555273253219,
    completed: 1555273253219,
    isPrivate: false,
    parentId: null,
    subtaskIds: [],
    subtaskNames: [],
    recurrence: { type: 'never' },
}

export const TasksDone_1 = {
    id: '-Txyz_3',
    done: true,
    name: 'My new done task',
    userIds: ['-Uxyz_1'],
    hasStar: false,
    created: 1555192800000,
    creatorId: '-Uxyz_1',
    dueDate: 1555273253219,
    completed: 1555273253219,
    isPrivate: false,
    parentId: null,
    subtaskIds: [],
    subtaskNames: [],
    recurrence: { type: 'never' },
}

export const TasksListInput_1 = [TasksOpen_1, TasksPending_1, TasksDone_1]
