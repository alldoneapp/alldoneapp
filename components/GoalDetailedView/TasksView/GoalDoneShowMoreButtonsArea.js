import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { View, StyleSheet } from 'react-native'

import ShowMoreButton from '../../UIControls/ShowMoreButton'
import { setGoalDoneTasksExpandedAmount, setShowLimitedFeatureModal } from '../../../redux/actions'
import { AMOUNT_OF_EARLIER_TASKS_TO_SHOW_WHEN_PRESS_BUTTON } from '../../../utils/backends/doneTasks'
import { PLAN_STATUS_PREMIUM } from '../../Premium/PremiumHelper'

export default function GoalDoneShowMoreButtonsArea() {
    const dispatch = useDispatch()
    const goalDoneTasksData = useSelector(state => state.goalDoneTasksData)
    const goalDoneTasksExpandedAmount = useSelector(state => state.goalDoneTasksExpandedAmount)
    const premiumStatus = useSelector(state => state.loggedUser.premium.status)

    const expandTasks = () => {
        premiumStatus === PLAN_STATUS_PREMIUM
            ? dispatch(
                  setGoalDoneTasksExpandedAmount(
                      goalDoneTasksExpandedAmount + AMOUNT_OF_EARLIER_TASKS_TO_SHOW_WHEN_PRESS_BUTTON
                  )
              )
            : dispatch(setShowLimitedFeatureModal(true))
    }

    const contractTasks = () => {
        dispatch(
            setGoalDoneTasksExpandedAmount(
                goalDoneTasksExpandedAmount - AMOUNT_OF_EARLIER_TASKS_TO_SHOW_WHEN_PRESS_BUTTON
            )
        )
    }

    const getTotalTasksAmount = () => {
        let amount = 0
        goalDoneTasksData.forEach(taskData => {
            amount += taskData[1]
        })
        return amount
    }

    const firstDayTasksAmount = goalDoneTasksData.length > 0 ? goalDoneTasksData[0][1] : 0
    const totalTasksAmount = getTotalTasksAmount()

    const showEarlierTasksButton = totalTasksAmount > firstDayTasksAmount + goalDoneTasksExpandedAmount
    const showHideEarlierTasksButton = goalDoneTasksExpandedAmount > 0

    return (
        <View style={localStyles.container}>
            {showEarlierTasksButton && (
                <ShowMoreButton
                    expanded={false}
                    expand={expandTasks}
                    expandText={'earlier tasks'}
                    style={{ flex: 0, marginRight: 16 }}
                />
            )}
            {showHideEarlierTasksButton > 0 && (
                <ShowMoreButton
                    expanded={true}
                    contract={contractTasks}
                    contractText={'hide earlier tasks'}
                    style={{ flex: 0 }}
                />
            )}
        </View>
    )
}

const localStyles = StyleSheet.create({
    container: {
        flexDirection: 'row',
        justifyContent: 'center',
    },
})
