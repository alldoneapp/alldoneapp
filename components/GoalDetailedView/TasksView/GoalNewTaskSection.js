import React, { useRef, useEffect } from 'react'
import moment from 'moment'
import { StyleSheet, View } from 'react-native'
import { useSelector, useDispatch } from 'react-redux'

import DismissibleItem from '../../UIComponents/DismissibleItem'
import store from '../../../redux/store'
import { BACKLOG_DATE_NUMERIC, BACKLOG_DATE_STRING } from '../../TaskListView/Utils/TasksHelper'
import { isInputsFocused } from '../../../utils/HelperFunctions'
import AddTask from '../../TaskListView/AddTask'
import { GOAL_OPEN_TASKS_EXPANDED_LATER_DAYS, GOAL_OPEN_TASKS_EXPANDED_SOMEDAY } from '../../GoalsView/GoalsHelper'
import { setGoalOpenMainTasksExpanded, setGoalOpenTasksExpandState } from '../../../redux/actions'
import { DATE_TASK_INDEX, MAIN_TASK_INDEX } from '../../../utils/backends/Tasks/openGoalTasks'
import ProjectHelper from '../../SettingsView/ProjectsSettings/ProjectHelper'
import { PROJECT_TYPE_TEMPLATE } from '../../SettingsView/ProjectsSettings/ProjectsSettings'
import EditTask from '../../TaskListView/TaskItem/EditTask'

export default function GoalNewTaskSection({ projectId, goal, dateFormated }) {
    const dispatch = useDispatch()
    const newItemRef = useRef(null)
    const goalOpenTasksData = useSelector(state => state.goalOpenTasksData)
    const goalOpenTasksExpandState = useSelector(state => state.goalOpenTasksExpandState)

    const date = moment(dateFormated, 'YYYYMMDD')

    const onKeyDown = e => {
        if (store.getState().blockShortcuts) {
            return
        }
        const { lastAddNewTaskDate } = store.getState()
        const { projectId: lastPId, date: lastAddDate } = lastAddNewTaskDate
            ? lastAddNewTaskDate
            : { projectId: null, date: null }

        const shouldOpen =
            (lastAddDate == null && projectId === lastPId && date.isSame(moment(), 'day')) ||
            (lastAddDate != null && projectId === lastPId && date.isSame(moment(lastAddDate), 'day'))

        const dismissItems = document.querySelectorAll('[aria-label="dismissible-edit-item"]')
        if (e.key === '+' && dismissItems.length === 0 && !isInputsFocused() && shouldOpen) {
            e.preventDefault()
            e.stopPropagation()
            newItemRef.current.toggleModal()
        }
    }

    const tryExpandTasksListInGoalWhenAddTask = task => {
        const { dueDate } = task
        if (goalOpenTasksExpandState !== GOAL_OPEN_TASKS_EXPANDED_SOMEDAY) {
            const taskInBacklog = dueDate === BACKLOG_DATE_NUMERIC
            if (taskInBacklog) {
                dispatch(setGoalOpenTasksExpandState(GOAL_OPEN_TASKS_EXPANDED_SOMEDAY))
            } else if (
                goalOpenTasksData.length > 0 &&
                goalOpenTasksExpandState !== GOAL_OPEN_TASKS_EXPANDED_LATER_DAYS &&
                goalOpenTasksData[0][DATE_TASK_INDEX] !== moment(dueDate).format('YYYYMMDD')
            ) {
                dispatch(setGoalOpenTasksExpandState(GOAL_OPEN_TASKS_EXPANDED_LATER_DAYS))
            }
        }

        const { loggedUser } = store.getState()

        if (
            goalOpenTasksData.length > 0 &&
            goalOpenTasksData[0][DATE_TASK_INDEX] === moment(dueDate).format('YYYYMMDD') &&
            loggedUser.numberTodayTasks > 0 &&
            loggedUser.numberTodayTasks < goalOpenTasksData[0][MAIN_TASK_INDEX].length &&
            ProjectHelper.getTypeOfProject(loggedUser, projectId) === PROJECT_TYPE_TEMPLATE
        ) {
            dispatch(setGoalOpenMainTasksExpanded(true))
        }
    }

    useEffect(() => {
        document.addEventListener('keydown', onKeyDown)
        return () => {
            document.removeEventListener('keydown', onKeyDown)
        }
    })

    return (
        <View style={localStyles.container}>
            <DismissibleItem
                ref={newItemRef}
                defaultComponent={
                    <AddTask
                        projectId={projectId}
                        tags={[]}
                        toggleModal={() => {
                            newItemRef?.current?.toggleModal()
                        }}
                        newItem={true}
                        hideParentGoalButton={true}
                        dateFormated={dateFormated}
                    />
                }
                modalComponent={
                    <EditTask
                        adding={true}
                        projectId={projectId}
                        onCancelAction={() => {
                            newItemRef?.current?.toggleModal()
                        }}
                        defaultDate={date.valueOf()}
                        inBacklog={dateFormated === BACKLOG_DATE_STRING}
                        activeGoal={goal}
                        tryExpandTasksListInGoalWhenAddTask={tryExpandTasksListInGoalWhenAddTask}
                        dateFormated={dateFormated}
                        useLoggedUser={true}
                    />
                }
            />
        </View>
    )
}

const localStyles = StyleSheet.create({
    container: {
        paddingHorizontal: 8,
    },
})
