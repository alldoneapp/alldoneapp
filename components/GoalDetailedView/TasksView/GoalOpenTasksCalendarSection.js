import React from 'react'
import { StyleSheet, Text, TouchableOpacity, View } from 'react-native'

import styles, { colors } from '../../styles/global'
import GoogleCalendar from '../../../assets/svg/GoogleCalendar'
import { CALENDAR_TASK_INDEX } from '../../../utils/backends/Tasks/openGoalTasks'
import GoalTasksList from './GoalTasksList'

export default function GoalOpenTasksCalendarSection({ projectId, calendarTasks, dateIndex, isActiveOrganizeMode }) {
    const openLink = () => {
        return window.open(
            'https://calendar.google.com/calendar/u/?' + `authuser=${calendarTasks[0].calendarData.email}`,
            '_blank'
        )
    }

    return (
        <View style={localStyles.container}>
            <View style={localStyles.subContainer}>
                <View style={localStyles.centeredRow}>
                    <TouchableOpacity onPress={openLink} style={{ flexDirection: 'row' }}>
                        <GoogleCalendar />
                        <Text style={localStyles.title}>Google Calendar</Text>
                    </TouchableOpacity>
                </View>
            </View>

            <GoalTasksList
                projectId={projectId}
                taskList={calendarTasks}
                dateIndex={dateIndex}
                taskListIndex={CALENDAR_TASK_INDEX}
                isActiveOrganizeMode={isActiveOrganizeMode}
            />
        </View>
    )
}

const localStyles = StyleSheet.create({
    container: {
        flexDirection: 'column',
    },
    subContainer: {
        flexDirection: 'row',
        alignItems: 'center',
        height: 48,
        marginTop: 52,
        paddingBottom: 2,
        paddingLeft: 2,
    },
    centeredRow: {
        flex: 1,
        maxHeight: 28,
        flexDirection: 'row',
        alignItems: 'center',
    },
    title: {
        ...styles.caption1,
        color: colors.Text03,
        marginLeft: 8,
    },
})
