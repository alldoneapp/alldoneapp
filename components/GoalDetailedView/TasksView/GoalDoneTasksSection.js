import React, { useEffect } from 'react'
import { StyleSheet, View } from 'react-native'
import { useSelector, useDispatch } from 'react-redux'

import DoneTasksByDate from '../../TaskListView/DoneTasksView/DoneTasksByDate'
import GoalDoneShowMoreButtonsArea from './GoalDoneShowMoreButtonsArea'
import { setGoalDoneTasksExpandedAmount } from '../../../redux/actions'
import { taskMatchHashtagFilters } from '../../HashtagFilters/FilterHelpers/FilterTasks'
import { AMOUNT_OF_EARLIER_TASKS_TO_SHOW_WHEN_PRESS_BUTTON } from '../../../utils/backends/doneTasks'

export default function GoalDoneTasksSection({ projectId }) {
    const dispatch = useDispatch()
    const goalDoneTasksData = useSelector(state => state.goalDoneTasksData)
    const goalDoneSubtasksByParent = useSelector(state => state.goalDoneSubtasksByParent)
    const goalDoneTasksExpandedAmount = useSelector(state => state.goalDoneTasksExpandedAmount)
    const isAnonymous = useSelector(state => state.loggedUser.isAnonymous)
    const hashtagFilters = useSelector(state => state.hashtagFilters) //NEEDED FOR FORCE A RENDER

    useEffect(() => {
        const amountFirstDay = goalDoneTasksData.length > 0 ? goalDoneTasksData[0][1] : 0
        let amount = -amountFirstDay
        goalDoneTasksData.forEach(taskData => {
            amount += taskData[1]
        })

        dispatch(
            setGoalDoneTasksExpandedAmount(
                Math.ceil(amount / AMOUNT_OF_EARLIER_TASKS_TO_SHOW_WHEN_PRESS_BUTTON) *
                    AMOUNT_OF_EARLIER_TASKS_TO_SHOW_WHEN_PRESS_BUTTON
            )
        )
        return () => {
            dispatch(setGoalDoneTasksExpandedAmount(0))
        }
    }, [])

    let tasksLeftToShow = goalDoneTasksExpandedAmount + (goalDoneTasksData.length > 0 ? goalDoneTasksData[0][1] : 0)

    return (
        <View style={localStyles.container}>
            {goalDoneTasksData.map((tasksData, index) => {
                const tasksList = tasksData[3].slice(0, tasksLeftToShow)
                tasksLeftToShow -= tasksList.length

                return (
                    <DoneTasksByDate
                        key={tasksData[0]}
                        projectId={projectId}
                        taskList={tasksList.filter(item => taskMatchHashtagFilters(item))}
                        dateFormated={tasksData[0]}
                        firstDateSection={index === 0}
                        subtaskByTask={goalDoneSubtasksByParent}
                        estimation={tasksData[2]}
                    />
                )
            })}
            {!isAnonymous && goalDoneTasksData.length > 0 && <GoalDoneShowMoreButtonsArea />}
        </View>
    )
}

const localStyles = StyleSheet.create({
    container: {
        marginBottom: 16,
    },
})
