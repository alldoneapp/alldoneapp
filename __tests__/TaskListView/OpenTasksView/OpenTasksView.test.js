/**
 * @jest-environment jsdom
 */

import React from 'react'
import OpenTasksView from '../../../components/TaskListView/OpenTasksView/OpenTasksView'
import { OpenTasksViewInput } from '../../../__mocks__/MockData/TasksView/OpenTasksViewInput'

import renderer from 'react-test-renderer'
import store from '../../../redux/store'
import { setAmountTasksByProjects, storeCurrentUser, storeLoggedUserProjects } from '../../../redux/actions'
import { DefaultAmountTasks } from '../../../__mocks__/MockData/TasksView/DefaultAmountTasks'

const userId = 'C08CK8x1I5YS2lxVixuLHaF3SrA3'
let loggedUserProjects = [
    { id: '-Asd', name: 'My project', color: '#0055ff' },
    { id: '-Asd2', name: 'My project 2', color: '#0055ff' },
]
let currentUser = { uid: userId, photoURL: 'http:path.to.photo', displayName: 'Chicho' }

describe('OpenTasksView component', () => {
    beforeEach(() => {
        store.dispatch([
            storeLoggedUserProjects(loggedUserProjects),
            storeCurrentUser(currentUser),
            store.dispatch(setAmountTasksByProjects(DefaultAmountTasks)),
        ])
    })

    describe('OpenTasksView snapshot test', () => {
        xit('should render correctly', async () => {
            const tree = renderer.create(<OpenTasksView taskList={OpenTasksViewInput} />).toJSON()
            expect(tree).toMatchSnapshot()
        })
    })

    describe('Function componentWillUnmount snapshot test', () => {
        xit('Unmount function should execute correctly', () => {
            const tree = renderer.create(<OpenTasksView taskList={OpenTasksViewInput} />)
            tree.unmount()
        })
    })

    describe('Function updateState snapshot test', () => {
        xit('should execute and render correctly', () => {
            const tree = renderer.create(<OpenTasksView taskList={OpenTasksViewInput} />)
            expect(tree.toJSON()).toMatchSnapshot()

            tree.getInstance().updateState()
            expect(tree.toJSON()).toMatchSnapshot()
        })
    })
})
