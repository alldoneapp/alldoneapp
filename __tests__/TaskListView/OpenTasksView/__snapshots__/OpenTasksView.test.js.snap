// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OpenTasksView component Function updateState snapshot test should execute and render correctly 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "white",
        "paddingHorizontal": 56,
      },
      undefined,
    ]
  }
/>
`;

exports[`OpenTasksView component Function updateState snapshot test should execute and render correctly 2`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "white",
        "paddingHorizontal": 56,
      },
      undefined,
    ]
  }
/>
`;

exports[`OpenTasksView component OpenTasksView snapshot test should render correctly 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "white",
        "paddingHorizontal": 56,
      },
      undefined,
    ]
  }
/>
`;
